_posts/2016-04-02-xss-justthefacts.md

```diff
@@ -0,0 +1,48 @@
+---
+layout:     post
+title:      XSS - Just the facts, m'aam
+author:     dan_adams
+date:       2016-04-02 11:00:00
+summary:    A basic primer on Cross-site scripting (CSS) and its prevention
+category:   security
+tags:       security scripting xss sanitization encoding vulnerability exploit
+---
+
+One of the simplest and most prevalent vulnerabilities found on web applications that we pentest is Cross-Site Scripting or XSS. It is still the third most critical vulnerability according to the [https://www.owasp.org/index.php/Top10](Owasp Top 10 Vulnerabilities) list as of 2013, and closely linked to the #1 most critical vulnerability, Injection.
+
+Clearly, despite its ubiquity, there is still a lot of uncertainty amongst developers and others around what exactly XSS is and how to prevent it. Very often the topic is made overly complex, when in reality it is relatively simple to both understand and defend against.
+
+Cross-site scripting is the slightly clunky name for the malicious insertion of unauthorised javascript into a website or web application. Javascript can be dangerous because it executes client-side within the context of the end user's browser and hence can have access to customer data as well as control of browser actions. Malicious payloads can include:
+* Capturing user input such as passwords via a keylogger
+* Sending cookies, tokens and other cached data to a third party
+* Performing network requests and system operations that the user hasn't requested
+* Forcing downloads of files to the end user PC
+
+At its simplest, XSS relies on a website returning executable code to a client that a third party has injected, rather than data that the webserver is intending to serve. This can take a variety of forms but usually relies on the website returning a parameter to a customer that the customer has previously entered in an earlier web request. That is, it relies on data that is intended to be passive/handled data (such as raw text) to transform context into some form of exectuable payload such as a command or script, which can then execute. If the parameter containing the executable code is passed to and executes on the server side, it can lead to a SQL injection or command execution attack, and if executed on the client side, it can lead to an XSS attack.
+
+An example of a website that could be vulnerable to this could be as simple as a site containing a single form that echoes your name back to you: you enter your name ("Bob") in a login form field, and the website returns "Hello, Bob" in its HTTP response. If the website can instead be forced to process the name of "&lt;script&amp;... &lt;script&amp;" then it is possible to inject javascript into unsuspecting visitors' browsers.
+
+![Bobby Tables cartoon](https://imgs.xkcd.com/comics/exploits_of_a_mom.png)
```

Comment:
I love the cartoon, and it's at the contextually correct place, but it's SQL injection not XSS ;(